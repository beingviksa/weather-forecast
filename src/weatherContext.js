import { createContext, useContext, useState } from "react";

export const WeatherContext = createContext();

export const WeatherContextProvider = ({ children }) => {
  const [searchResults, setSearchResults] = useState([]);
  const [consolidatedData, setConsolidatedData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");
  return (
    <WeatherContext.Provider
      value={{
        searchResults,
        setSearchResults,
        consolidatedData,
        setConsolidatedData,
        isLoading,
        setIsLoading,
        error,
        setError,
      }}>
      {children}
    </WeatherContext.Provider>
  );
};

export const useWeatherContext = () => {
  return useContext(WeatherContext);
};
