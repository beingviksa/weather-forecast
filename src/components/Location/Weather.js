import { useState, useEffect } from "react";
import axios from "axios";
import { useWeatherContext } from "../../weatherContext";
import TodaysWeather from "../TodaysWeather";
import HourlyWeather from "../HourlyWeather";
import SearchBox from "../SearchBox";
import Error from "../Error";
import Loading from "../Loading";
import {
  getTodaysMinMaxTemp,
  getConsolidatedTempData,
} from "../../utills/helper";

const Weather = ({ city }) => {
  const { setConsolidatedData, isLoading, setIsLoading, error, setError } =
    useWeatherContext();
  const [minMaxTemp, setMinMaxTemp] = useState({ min: "", max: "" });

  useEffect(() => {
    const fetchData = async () => {
      try {
        setIsLoading(true);
        const res = await axios.get(
          `https://api.open-meteo.com/v1/forecast?latitude=${city.coord.lat}&longitude=${city.coord.lon}&hourly=temperature_2m,relativehumidity_2m,rain`
        );
        setIsLoading(false);
        const weatherData = res.data;
        if (weatherData) {
          getHourlyWeather(weatherData?.hourly);
        }
      } catch (error) {
        setIsLoading(false);
        setError(error.message);
      }
    };
    fetchData();
  }, [city]);

  const getHourlyWeather = (hourlyData) => {
    // filterd 24 hours data
    const timeData = hourlyData?.time.slice(0, 24);
    const temperatureData = hourlyData?.temperature_2m.slice(0, 24);
    const humidityData = hourlyData?.relativehumidity_2m.slice(0, 24);
    const rainData = hourlyData?.rain.slice(0, 24);

    const hourlyWeatherInfo = {
      timeData,
      temperatureData,
      humidityData,
      rainData,
    };

    const todaysTemp = getTodaysMinMaxTemp(temperatureData);
    if (todaysTemp) {
      setMinMaxTemp({
        min: todaysTemp.minTemp.toFixed(),
        max: todaysTemp.maxTemp.toFixed(),
      });
    }

    if (!!hourlyWeatherInfo.timeData) {
      const data = getConsolidatedTempData(hourlyWeatherInfo);
      setConsolidatedData(data);
    }
  };

  if (!!error) {
    return <Error error={error} />;
  }
  if (isLoading) {
    return <Loading />;
  }

  return (
    <div className="container">
      <SearchBox showBackButton />
      <TodaysWeather city={city} minMaxTemp={minMaxTemp} />
      <HourlyWeather />
    </div>
  );
};

export default Weather;
