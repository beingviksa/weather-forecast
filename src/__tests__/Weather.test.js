import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import axios from "axios";
import Weather from "../components/Location/Weather";
import { WeatherContext } from "../weatherContext";

afterEach(() => {
  axios.get.mockClear();
});

describe("<Weather/>", () => {
  const props = {
    city: {
      name: "Paris",
      country: "FR",
      id: 2968815,
      coord: {
        lat: 48.853401,
        lon: 2.3486,
      },
    },
  };
  const contextData = {
    setSearchResults: jest.fn(),
    isLoading: false,
    setIsLoading: jest.fn(),
    error: "",
    setError: jest.fn(),
  };
  axios.get.mockResolvedValueOnce({
    data: {
      latitude: 48.86,
      longitude: 2.3399997,
      generationtime_ms: 0.6700754165649414,
      utc_offset_seconds: 0,
      timezone: "GMT",
      timezone_abbreviation: "GMT",
      elevation: 43,
      hourly_units: {
        time: "iso8601",
        temperature_2m: "Â°C",
        relativehumidity_2m: "%",
        rain: "mm",
      },
      hourly: {
        time: [
          "2022-10-12T00:00",
          "2022-10-12T01:00",
          "2022-10-12T02:00",
          "2022-10-12T03:00",
          "2022-10-12T04:00",
          "2022-10-12T05:00",
          "2022-10-12T06:00",
        ],
        temperature_2m: [9, 8.7, 8.4, 7.4, 7, 6.6, 6.4],
        relativehumidity_2m: [85, 85, 85, 88, 89, 91, 93],
        rain: [0, 0, 0, 0, 0, 0, 0],
      },
    },
  });
  it("render the Weather Componenent", async () => {
    const { container } = render(
      <WeatherContext.Provider value={contextData}>
        <Router>
          <Weather {...props} />
        </Router>
      </WeatherContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("show loading spinner when it is fetching data", () => {
    render(
      <WeatherContext.Provider value={contextData}>
        <Router>
          <Weather {...props} />
        </Router>
      </WeatherContext.Provider>
    );

    expect(contextData.setIsLoading).toHaveBeenCalledWith(true);
  });

  it("fetch the weather data as per the selected city", async () => {
    const { lat, lon } = props.city.coord;

    render(
      <WeatherContext.Provider value={contextData}>
        <Router>
          <Weather {...props} />
        </Router>
      </WeatherContext.Provider>
    );

    const url = `https://api.open-meteo.com/v1/forecast?latitude=${lat}&longitude=${lon}&hourly=temperature_2m,relativehumidity_2m,rain`;

    expect(axios.get).toHaveBeenCalledWith(url);
    expect(axios.get).toHaveBeenCalledTimes(1);
  });
});
