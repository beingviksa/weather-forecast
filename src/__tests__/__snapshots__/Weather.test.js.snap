// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather/> render the Weather Componenent 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="search"
    >
      <a
        class="back-link"
        href="/"
      >
        ← Home
      </a>
      <input
        data-testid="search-box"
        placeholder="Seach for a city..."
        type="text"
        value=""
      />
    </div>
    <div
      class="todays-forecast"
    >
      <h1
        data-testid="city-country"
      >
        Paris
         
        (FR)
      </h1>
      <section>
        <div
          class="temp-details"
        >
          <span
            class="temp-label"
          >
            Min.
          </span>
          <span
            class="min-temp"
            data-testid="min-temp"
          >
            °C
          </span>
        </div>
        <div
          class="temp-details"
        >
          <span
            class="temp-label"
          >
            Max.
          </span>
          <span
            class="max-temp"
            data-testid="max-temp"
          >
            °C
          </span>
        </div>
      </section>
    </div>
    <h1>
      ...Loading
    </h1>
  </div>
</div>
`;
